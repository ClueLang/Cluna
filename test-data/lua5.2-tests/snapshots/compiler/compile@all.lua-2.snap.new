---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
