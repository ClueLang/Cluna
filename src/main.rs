#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
