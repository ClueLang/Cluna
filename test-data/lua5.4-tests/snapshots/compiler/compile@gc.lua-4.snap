---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.4-tests/gc.lua
---
print('testing incremental garbage collection');
local debug = require("debug")
assert(collectgarbage("isrunning"));
collectgarbage();
local oldmode = collectgarbage("incremental")
assert(collectgarbage("generational") == "incremental");
assert(collectgarbage("generational") == "generational");
assert(collectgarbage("incremental") == "generational");
assert(collectgarbage("incremental") == "incremental");
local fn nop() {
    
}
local fn gcinfo() {
    return collectgarbage("count") * 1024
}
{
    local a = collectgarbage("setpause", 200)
    local b = collectgarbage("setstepmul", 200)
    local t = {
        0, 
        2, 
        10, 
        90, 
        500, 
        5000, 
        30000, 
        0x7ffffffe
    }
    for i = 1, #t {
        local p = t[(i)]
        for j = 1, #t {
            local m = t[(j)]
            collectgarbage("setpause", p);
            collectgarbage("setstepmul", m);
            collectgarbage("step", 0);
            collectgarbage("step", 10000);
        }
    }
    collectgarbage("setpause", a);
    collectgarbage("setstepmul", b);
    collectgarbage();
}
_G[("while")] = 234
local fn GC1() {
    local u
    local b
    local finish = false
    u = setmetatable({}, {
        __gc = fn () {
            finish = true
        }
    })
    b = {
        34
    }
    loop  {
        u = {}
    } until finish
    assert(b[(1)] == 34);
    finish = false
    local i = 1
    u = setmetatable({}, {
        __gc = fn () {
            finish = true
        }
    })
    loop  {
        i = i + 1
        u = tostring(i) .. tostring(i)
    } until finish
    assert(b[(1)] == 34);
    finish = false
    u = setmetatable({}, {
        __gc = fn () {
            finish = true
        }
    })
    loop  {
        local i
        u = fn () {
            return i
        }
    } until finish
    assert(b[(1)] == 34);
}
local fn GC2() {
    local u
    local finish = false
    u = {
        setmetatable({}, {
            __gc = fn () {
                finish = true
            }
        })
    }
    local b = {
        34
    }
    loop  {
        u = {
            {}
        }
    } until finish
    assert(b[(1)] == 34);
    finish = false
    local i = 1
    u = {
        setmetatable({}, {
            __gc = fn () {
                finish = true
            }
        })
    }
    loop  {
        i = i + 1
        u = {
            tostring(i) .. tostring(i)
        }
    } until finish
    assert(b[(1)] == 34);
    finish = false
    u = {
        setmetatable({}, {
            __gc = fn () {
                finish = true
            }
        })
    }
    loop  {
        local i
        u = {
            fn () {
                return i
            }
        }
    } until finish
    assert(b[(1)] == 34);
}
local fn GC() {
    GC1();
    GC2();
}
{
    print("creating many objects");
    local limit = 5000
    for i = 1, limit {
        local a = {}
        a = nil
    }
    local a = "a"
    for i = 1, limit {
        a = i .. "b"
        a = string.gsub(a, '(%d%d*)', "%1 %1")
        a = "a"
    }
    a = {}
    method a::test() {
        for i = 1, limit {
            load(string.format("function temp(a) return 'a%d' end", i), "")();
            assert(temp() == string.format('a%d', i));
        }
    }
    a::test();
}
{
    local f = fn () {
        
    }
}
print("functions with errors");
prog = `
do
  a = 10;
  function foo(x,y)
    a = sin(a+0.456-0.23e-12);
    return function (z) return sin(%x+z) end
  end
  local x = function (w) a=a+w; end
end
`
{
    local step = 1
    if _soft {
        step = 13
    }
    for i = 1, string.len(prog), step {
        for j = i, string.len(prog), step {
            pcall(load(string.sub(prog, i, j), ""));
        }
    }
}
foo = nil
print('long strings');
x = "01234567890123456789012345678901234567890123456789012345678901234567890123456789"
assert(string.len(x) == 80);
s = ''
k = math.min(300, (math.maxinteger /_ 80) /_ 2)
for n = 1, k {
    s = s .. x
    j = tostring(n)
}
assert(string.len(s) == k * 80);
s = string.sub(s, 1, 10000)
s, i = string.gsub(s, '(%d%d%d%d)', '')
assert(i == 10000 /_ 4);
s = nil
x = nil
assert(_G[("while")] == 234);
{
    print("steps");
    print("steps (2)");
    local fn dosteps(siz) {
        collectgarbage();
        local a = {}
        for i = 1, 100 {
            a[(i)] = {
                {}
            }
            local b = {}
        }
        local x = gcinfo()
        local i = 0
        loop  {
            i = i + 1
        } until collectgarbage("step", siz)
        assert(gcinfo() < x);
        return i
    }
    collectgarbage("stop");
    if !_port {
        assert(dosteps(10) < dosteps(2));
    }
    assert(dosteps(20000) == 1);
    assert(collectgarbage("step", 20000) == true);
    assert(collectgarbage("step", 20000) == true);
    assert(!collectgarbage("isrunning"));
    collectgarbage("restart");
    assert(collectgarbage("isrunning"));
}
if !_port {
    collectgarbage();
    collectgarbage();
    local x = gcinfo()
    collectgarbage("stop");
    loop  {
        local a = {}
    } until gcinfo() > 3 * x
    collectgarbage("restart");
    assert(collectgarbage("isrunning"));
    loop  {
        local a = {}
    } until gcinfo() <= x * 2
}
print("clearing tables");
lim = 15
a = {}
for i = 1, lim {
    a[({})] = i
}
b = {}
for k, v with pairs(a) {
    b[(k)] = v
}
for n with pairs(b) {
    a[(n)] = undef
    assert(type(n) == 'table' && next(n) == nil);
    collectgarbage();
}
b = nil
collectgarbage();
for n with pairs(a) {
    error('cannot be here');
}
for i = 1, lim {
    a[(i)] = i
}
for i = 1, lim {
    assert(a[(i)] == i);
}
print('weak tables');
a = {}
setmetatable(a, {
    __mode = 'k'
});
for i = 1, lim {
    a[({})] = i
}
for i = 1, lim {
    a[(i)] = i
}
for i = 1, lim {
    local s = string.rep('@', i)
    a[(s)] = s .. '#'
}
collectgarbage();
local i = 0
for k, v with pairs(a) {
    assert(k == v || k .. '#' == v);
    i = i + 1
}
assert(i == 2 * lim);
a = {}
setmetatable(a, {
    __mode = 'v'
});
a[(1)] = string.rep('b', 21)
collectgarbage();
assert(a[(1)]);
a[(1)] = undef
for i = 1, lim {
    a[(i)] = {}
}
for i = 1, lim {
    a[(i .. 'x')] = {}
}
for i = 1, lim {
    local t = {}
    a[(t)] = t
}
for i = 1, lim {
    a[(i + lim)] = i .. 'x'
}
collectgarbage();
local i = 0
for k, v with pairs(a) {
    assert(k == v || k - lim .. 'x' == v);
    i = i + 1
}
assert(i == 2 * lim);
a = {}
setmetatable(a, {
    __mode = 'kv'
});
local x, y, z = {}, {}, {}
a[(1)], a[(2)], a[(3)] = x, y, z
a[(string.rep('$', 11))] = string.rep('$', 11)
for i = 4, lim {
    a[(i)] = {}
}
for i = 1, lim {
    a[({})] = i
}
for i = 1, lim {
    local t = {}
    a[(t)] = t
}
collectgarbage();
assert(next(a) != nil);
local i = 0
for k, v with pairs(a) {
    assert((k == 1 && v == x) || (k == 2 && v == y) || (k == 3 && v == z) || k == v);
    i = i + 1
}
assert(i == 4);
x, y, z = nil
collectgarbage();
assert(next(a) == string.rep('$', 11));
a = {}
local t = {
    x = 10
}
local C = setmetatable({
    key = t
}, {
    __mode = 'v'
})
local C1 = setmetatable({
    t = 1
}, {
    __mode = 'k'
})
a.x = t
setmetatable(a, {
    __gc = fn (u) {
        assert(C.key == nil);
        assert(type(next(C1)) == 'table');
    }
});
a, t = nil
collectgarbage();
collectgarbage();
assert(next(C) == nil && next(C1) == nil);
C, C1 = nil
local mt = {
    __mode = 'k'
}
a = {
    {
        10
    }, 
    {
        20
    }, 
    {
        30
    }, 
    {
        40
    }
}
setmetatable(a, mt);
x = nil
for i = 1, 100 {
    local n = {}
    a[(n)] = {
        k = {
            x
        }
    }
    x = n
}
GC();
local n = x
local i = 0
while n {
    n = a[(n)].k[(1)]
    i = i + 1
}
assert(i == 100);
x = nil
GC();
for i = 1, 4 {
    assert(a[(i)][(1)] == i * 10);
    a[(i)] = undef
}
assert(next(a) == nil);
local K = {}
a[(K)] = {}
for i = 1, 10 {
    a[(K)][(i)] = {}
    a[(a[(K)][(i)])] = setmetatable({}, mt)
}
x = nil
local k = 1
for j = 1, 100 {
    local n = {}
    local nk = k % 10 + 1
    a[(a[(K)][(nk)])][(n)] = {
        x, 
        k = k
    }
    x = n
    k = nk
}
GC();
local n = x
local i = 0
while n {
    local t = a[(a[(K)][(k)])][(n)]
    n = t[(1)]
    k = t.k
    i = i + 1
}
assert(i == 100);
K = nil
GC();
if T {
    collectgarbage("stop");
    local u = {}
    local s = {}
    setmetatable(s, {
        __mode = 'k'
    });
    setmetatable(u, {
        __gc = fn (o) {
            local i = s[(o)]
            s[(i)] = true
            assert(!s[(i - 1)]);
            if i == 8 {
                error("@expected@");
            }
        }
    });
    for i = 6, 10 {
        local n = setmetatable({}, getmetatable(u))
        s[(n)] = i
    }
    warn("@on");
    warn("@store");
    collectgarbage();
    assert(string.find(_WARN, "error in __gc"));
    assert(string.match(_WARN, "@(.-)@") == "expected");
    _WARN = false
    for i = 8, 10 {
        assert(s[(i)]);
    }
    for i = 1, 5 {
        local n = setmetatable({}, getmetatable(u))
        s[(n)] = i
    }
    collectgarbage();
    for i = 1, 10 {
        assert(s[(i)]);
    }
    getmetatable(u).__gc = nil
    warn("@normal");
}
print('+');
if T == nil {
    (Message || print)('\n >>> testC not active: skipping userdata GC tests <<<\n');
} else {
    local fn newproxy(u) {
        return debug.setmetatable(T.newuserdata(0), debug.getmetatable(u))
    }
    collectgarbage("stop");
    local u = newproxy(nil)
    debug.setmetatable(u, {
        __gc = true
    });
    local s = 0
    local a = {
        u = 0
    }
    setmetatable(a, {
        __mode = 'vk'
    });
    for i = 1, 10 {
        a[(newproxy(u))] = i
    }
    for k with pairs(a) {
        assert(getmetatable(k) == getmetatable(u));
    }
    local a1 = {}
    for k, v with pairs(a) {
        a1[(k)] = v
    }
    for k, v with pairs(a1) {
        a[(v)] = k
    }
    for i = 1, 10 {
        assert(a[(i)]);
    }
    getmetatable(u).a = a1
    getmetatable(u).u = u
    {
        local u = u
        getmetatable(u).__gc = fn (o) {
            assert(a[(o)] == 10 - s);
            assert(a[(10 - s)] == undef);
            assert(getmetatable(o) == getmetatable(u));
            assert(getmetatable(o).a[(o)] == 10 - s);
            s = s + 1
        }
    }
    a1, u = nil
    assert(next(a) != nil);
    collectgarbage();
    assert(s == 11);
    collectgarbage();
    assert(next(a) == nil);
}
local u = setmetatable({}, {
    __gc = true
})
setmetatable(getmetatable(u), {
    __mode = "v"
});
getmetatable(u).__gc = fn (o) {
    os.exit(1);
}
u = nil
collectgarbage();
local u = setmetatable({}, {
    __gc = true
})
local m = getmetatable(u)
m.x = {
    {
        0
    } = 1, 
    0 = {
        1
    }
}
setmetatable(m.x, {
    __mode = "kv"
});
m.__gc = fn (o) {
    assert(next(getmetatable(o).x) == nil);
    m = 10
}
u, m = nil
collectgarbage();
assert(m == 10);
{
    collectgarbage();
    collectgarbage();
    local m = collectgarbage("count")
    local a = setmetatable({}, {
        __mode = "kv"
    })
    a[(string.rep("a", 2 ^ 22))] = 25
    a[(string.rep("b", 2 ^ 22))] = {}
    a[({})] = 14
    assert(collectgarbage("count") > m + 2 ^ 13);
    collectgarbage();
    assert(collectgarbage("count") >= m + 2 ^ 12 && collectgarbage("count") < m + 2 ^ 13);
    local k, v = next(a)
    assert(k == string.rep("a", 2 ^ 22) && v == 25);
    assert(next(a, k) == nil);
    assert(a[(string.rep("b", 2 ^ 22))] == undef);
    a[(k)] = undef
    k = nil
    collectgarbage();
    assert(next(a) == nil);
    assert(a[(string.rep("b", 100))] == undef);
    assert(collectgarbage("count") <= m + 1);
}
if T {
    warn("@store");
    u = setmetatable({}, {
        __gc = fn () {
            error("@expected error");
        }
    })
    u = nil
    collectgarbage();
    assert(string.find(_WARN, "@expected error"));
    _WARN = false
    warn("@normal");
}
if !_soft {
    print("long list");
    local a = {}
    for i = 1, 200000 {
        a = {
            next = a
        }
    }
    a = nil
    collectgarbage();
}
print("self-referenced threads");
local thread_id = 0
local threads = {}
local fn fn(thread) {
    local x = {}
    threads[(thread_id)] = fn () {
        thread = x
    }
    coroutine.yield();
}
while thread_id < 1000 {
    local thread = coroutine.create(fn)
    coroutine.resume(thread, thread);
    thread_id = thread_id + 1
}
{
    local collected = false
    collectgarbage();
    collectgarbage("stop");
    {
        local fn f(param) {
            (fn () {
                assert(type(f) == 'function' && type(param) == 'thread');
                param = {
                    param, 
                    f
                }
                setmetatable(param, {
                    __gc = fn () {
                        collected = true
                    }
                });
                coroutine.yield(100);
            })();
        }
        local co = coroutine.create(f)
        assert(coroutine.resume(co, co));
    }
    collectgarbage();
    assert(collected);
    collectgarbage("restart");
}
{
    collectgarbage();
    collectgarbage("stop");
    collectgarbage("step", 0);
    local x = gcinfo()
    loop  {
        for i = 1, 1000 {
            _ENV.a = {}
        }
    } until gcinfo() > 2 * x
    collectgarbage("restart");
}
if T {
    local fn foo() {
        local a = {
            x = 20
        }
        coroutine.yield(fn () {
            return a.x
        });
        assert(a.x == 20);
        a = {
            x = 30
        }
        assert(T.gccolor(a) == "white");
        coroutine.yield(100);
    }
    local t = setmetatable({}, {
        __mode = "kv"
    })
    collectgarbage();
    collectgarbage('stop');
    t.co = coroutine.wrap(foo)
    local f = t.co()
    T.gcstate("atomic");
    assert(T.gcstate() == "atomic");
    assert(t.co() == 100);
    assert(T.gccolor(t.co) == "white");
    T.gcstate("pause");
    assert(t.co == nil && f() == 30);
    collectgarbage("restart");
    local u = T.newuserdata(0, 1)
    collectgarbage();
    collectgarbage("stop");
    local a = {}
    T.gcstate("atomic");
    T.gcstate("sweepallgc");
    local x = {}
    assert(T.gccolor(u) == "black");
    assert(T.gccolor(x) == "white");
    debug.setuservalue(u, x);
    assert(T.gccolor(u) == "gray");
    collectgarbage("restart");
    print("+");
}
if T {
    local debug = require("debug")
    collectgarbage("stop");
    local x = T.newuserdata(0)
    local y = T.newuserdata(0)
    debug.setmetatable(y, {
        __gc = nop
    });
    debug.setmetatable(x, {
        __gc = nop
    });
    assert(T.gccolor(y) == "white");
    T.checkmemory();
    collectgarbage("restart");
}
if T {
    print("emergency collections");
    collectgarbage();
    collectgarbage();
    T.totalmem(T.totalmem() + 200);
    for i = 1, 200 {
        local a = {}
    }
    T.totalmem(0);
    collectgarbage();
    local t = T.totalmem("table")
    local a = {
        {}, 
        {}, 
        {}
    }
    assert(T.totalmem("table") == t + 4);
    t = T.totalmem("function")
    a = fn () {
        
    }
    assert(T.totalmem("function") == t + 1);
    t = T.totalmem("thread")
    a = coroutine.create(fn () {
        
    })
    assert(T.totalmem("thread") == t + 1);
}
{
    local setmetatable, assert, type, print, getmetatable = setmetatable, assert, type, print, getmetatable
    local tt = {}
    tt.__gc = fn (o) {
        assert(getmetatable(o) == tt);
        local a = 'xuxu' .. (10 + 3) .. 'joao', {}
        ___Glob = o
        setmetatable({}, tt);
        print(">>> closing state " .. "<<<\n");
    }
    local u = setmetatable({}, tt)
    ___Glob = {
        u
    }
}
if T {
    local error, assert, find, warn = error, assert, string.find, warn
    local n = 0
    local lastmsg
    local mt = {
        __gc = fn (o) {
            n = n + 1
            assert(n == o[(1)]);
            if n == 1 {
                _WARN = false
            } elseif if n == 2 {
                assert(find(_WARN, "@expected warning"));
                lastmsg = _WARN
            } else {
                assert(lastmsg == _WARN);
            }
            warn("@store");
            _WARN = false
            error("@expected warning");
        }
    }
    for i = 10, 1, -1 {
        table.insert(___Glob, setmetatable({
            i
        }, mt));
    }
}
assert(collectgarbage('isrunning'));
{
    local res = true
    setmetatable({}, {
        __gc = fn () {
            res = collectgarbage()
        }
    });
    collectgarbage();
    assert(!res);
}
collectgarbage(oldmode);
print('OK');
